encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/TypeToken.java

```diff
@@ -140,6 +141,34 @@ private ClassToken(Class<T> rawType, TypeTokenContainer typeArguments) {
     public TypeTokenContainer getTypeArguments() {
       return typeArguments;
     }
+
+    @Override
+    public int hashCode() {
+      return 11 * rawType.hashCode() + typeArguments.hashCode();
+    }
+
+    @Override
+    public boolean equals(@Nullable Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof ClassToken)) {
+        return false;
+      }
+      ClassToken<?> that = (ClassToken<?>) o;
+      return this.rawType.equals(that.rawType) && this.typeArguments.equals(that.typeArguments);
+    }
+
+    @NonNull
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("ClassToken{ ");
+      sb.append("RawType: ").append(rawType).append(", ");
+      sb.append("TypeArguments: ").append(typeArguments).append("");
+      sb.append(" }");
+      return sb.toString();
```

Comment:
Or to disambiguate, could it be something like `TypeToken{Optionsl<String>[]`?

Reply:
Great! I'll do that.

Reply:
It turns out that i can do `TypeToken{TypeToken{Optional<TypeToken{String}>}[]}` or `Optional<String>[]` without any efforts. However, If we want `TypeToken{Optional<String>[]}`, it would involve more code since TypeTokens are nested. wdyt? @vkryachko

Reply:
yeah, that's expected. But it seems that it's pretty straightforward to achieve, i.e. by adding an abstract method like `toStringRaw()` to the type token and have all subclasses implement it, while `toString()` itself is only needed in `TypeToken`, wdyt?

Reply:
Niceeeeee! I am impressed.